core/raftlease/store.go

```diff
@@ -198,24 +198,33 @@ func (s *Store) pinOp(operation string, key lease.Key, entity string, stop <-cha
 func (s *Store) Advance(duration time.Duration, stop <-chan struct{}) error {
 	s.prevTimeMu.Lock()
 	defer s.prevTimeMu.Unlock()
+
 	newTime := s.prevTime.Add(duration)
 	err := s.runOnLeader(&Command{
 		Version:   CommandVersion,
 		Operation: OperationSetTime,
 		OldTime:   s.prevTime,
 		NewTime:   newTime,
 	}, stop)
-	if globalclock.IsConcurrentUpdate(err) {
-		// Someone else updated before us - get the new time.
-		s.prevTime = s.fsm.GlobalTime()
-	} else if lease.IsTimeout(err) {
-		// Convert this to a globalclock timeout to match the Updater
-		// interface.
-		err = globalclock.ErrTimeout
-	} else if err == nil {
-		s.prevTime = newTime
+
+	if err != nil {
+		// If we timed out, convert the error to match the Updater interface.
+		if lease.IsTimeout(err) {
+			return globalclock.ErrTimeout
+		}
+
+		// If we had an incorrect notion of global time, just resync.
+		// The caller needs no awareness of this condition.
+		if globalclock.IsOutOfSyncUpdate(err) {
+			s.prevTime = s.fsm.GlobalTime()
```

Comment:
My problem with this is that you are calling Advance which might not actually advance the global time.
If you did:
```
s.prevTime = s.fsm.GlobalTime()
return s.Advance(duration)
```
or something along that line (you may not want to do exactly that, since it can generate infinite recursion).
But the concern is that we are saying "current global time is clearly advanced", but we didn't actually check.

Maybe:
```
newTime := s.fsm.GlobalTime()
if newTime - s.prevTime > duration:
  return nil
```

Just something to indicate that we have done the equivalent of what you asked for, rather than assuming that OutOfSync means it has already been done.

Reply:
This mirrors the prior behaviour, but I will look at what we can do here.

Reply:
Actually, when the worker became singular, *that* changed what this does. 

The clock updater worker moves its last update time forward **assuming that someone else ticked it forward**. Since there is no one else, this will have been resulting in effective time dilation whenever such an error was returned.

I don't think it's super important - accuracy is less important than consistency here, but we should consider that might need to attempt a 2 second increment if the penultimate one failed.

Reply:
The general design for the global clock is  "it needs to be monotonically increasing, no faster than real time"
So missing an update is "ok" in the general sense, but I'm concerned about having an API on a class which says "Advance" but the clock doesn't actually move forward.

Reply:
I've rectified this and continue to pass the out-of-sync up to the caller, who logs and retries appropriately.